src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/AvailableLocationsService.kt

```diff
@@ -0,0 +1,122 @@
+package uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service
+
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.client.activitiesappointments.ActivitiesAppointmentsClient
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.client.nomismapping.NomisMappingClient
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.client.prisonapi.PrisonApiClient
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.common.isTimesOverlap
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.config.PrisonRegime
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.Location
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.request.AvailableLocationsRequest
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.response.AvailableLocation
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.response.AvailableLocationsResponse
+import java.time.LocalDate
+import java.time.LocalTime
+
+@Service
+@Transactional(readOnly = true)
+class AvailableLocationsService(
+  private val locationsService: LocationsService,
+  private val bookedLocationsService: BookedLocationsService,
+  private val prisonRegime: PrisonRegime,
+) {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  /**
+   * This provides a snapshot (point in time) of available rooms at time of calling. When the user attempts to save the
+   * booking the room could have already been taken so may fail.
+   */
+  fun findAvailableLocations(request: AvailableLocationsRequest): AvailableLocationsResponse {
+    val startOfDay = prisonRegime.startOfDay(request.prisonCode!!)
+    val endOfDay = prisonRegime.endOfDay(request.prisonCode)
+    val prisonVideoLinkLocations = getDecoratedLocationsAt(request.prisonCode)
+    val bookedLocations = bookedLocationsService.findBooked(request.prisonCode, request.date!!, prisonVideoLinkLocations)
+    val duration = request.bookingDuration!!.toLong()
+
+    val availableLocations = buildList {
+      prisonVideoLinkLocations.forEach { location ->
+        // These time adjustments do not allow for PRE and POST meeting times.
+        var startTime = startOfDay
+        var endTime = startOfDay.plusMinutes(duration)
+
+        while (startTime.isBefore(endOfDay)) {
```

Comment:
As it stands this works in its basic form.  It will return available slots for the whole day though for each available room.  When you consider the jump is 15 mins each time there can be a lot of free rooms.

Still need to check against requested time slots which will slim it down and also against and decorated room constraints.

Still might want to consider a cap in terms of how many rooms we offer.

Reply:
Yes agreed. Maybe a 10 room / slot cap?